### ISSUE TITLE
Workers not bundling imported files for node environments

### ISSUE THREAD

**Issue by MartinDawson** [webpack/webpack#13179]
# Bug report

**What is the current behavior?**

Webpack outputs the workers correctly but is not bundling the included relative imports in those web workers.

**If the current behavior is a bug, please provide the steps to reproduce.**

webpack.config.js:
```js
const path = require("path");
const nodeExternals = require("webpack-node-externals");

module.exports = {
  mode: "production",
  entry: "./app.js",
  output: {
    path: path.join(__dirname, "build"),
    filename: "app.js",
  },
  target: "node",
  node: {
    __dirname: false,
    __filename: false,
  },
  externals: [nodeExternals()],
  module: {
    rules: [
      {
        exclude: /node_modules/,
      },
    ],
  },
};
```
app.js

```js
const Worker = require("worker_threads").Worker;

const worker = new Worker(new URL("./testWorker.js", import.meta.url), {
  type: "module",
});

worker.on("message", (msg) => {
  console.log(msg);
});
```

testWorker.js
```js
const {
  Worker,
  isMainThread,
  parentPort,
  workerData,
} = require("worker_threads");
const Hello = require("./hello");

parentPort.postMessage(Hello());
```

Hello.js

```js
const Hello = () => "HI";

module.exports = Hello;
```

Output:

![image](https://user-images.githubusercontent.com/15030491/115264696-71845e80-a12e-11eb-9280-93ddc50903ae.png)

You can see from the output that Hello.js is not being bundled when it should be, The workers themselves get bundled correctly but not any of the relative imports (or requires) in the worker files themselves. I've tried many different things including es6 imports and type module flag, worker-plugin etc but nothing will get the imports to be bundled.

**What is the expected behavior?**
The imports in the workers should be bundled in the build dir like in the source files. If anyone can suggest a work around for now even, that would be great. Thanks

**Other relevant information:**
webpack version:  5.33.2
Node.js version:  12.20.1
Operating System: Windows 10
Additional tools:

**Comment by alexander-akait** [comment]
In ` type: "module",` no `require`, it is ESM, also do not mix `import.meta.url` and `require`, it is bad

**Comment by MartinDawson** [comment]
@alexander-akait I tried with ES6 and type: module and still the imported files in the worker do not get bundled.

app.js:

```js
import { Worker } from "worker_threads";

const worker = new Worker(new URL("./testWorker.js", import.meta.url), {
  type: "module",
});

worker.on("message", (msg) => {
  console.log(msg);
});
```

Hello.js:

```js
const Hello = () => "HI";

export default Hello;
```

testWorker.js:

```js
import { Worker, isMainThread, parentPort, workerData } from "worker_threads";
import Hello from "./hello";

parentPort.postMessage(Hello());
```

Still, the Hello.js does not get bundled:

![image](https://user-images.githubusercontent.com/15030491/115267110-b6a99000-a130-11eb-8c58-f6da1732ea19.png)

**Comment by alexander-akait** [comment]
Because `Hello.js` is part of chunk, why do you expected extra file?

**Comment by alexander-akait** [comment]
You can use `splitChunks` and extra module(s) to separate chunk, but in case above it is unnecessary

**Comment by MartinDawson** [comment]
@alexander-akait Thanks for fast response. Maybe not an extra file then, but the file Hello.js is never loaded in the build at all.

**Comment by MartinDawson** [comment]
@alexander-akait Apologies, you are correct, it's working fine actually.

My real code is using comlink and other plugins and it wasn't bundling the chunk in. However this clearly works in the minimum example above so I will close it as it does not seem to be a webpack issue.

Thanks

**Comment by alexander-akait** [comment]
> Maybe not an extra file then, but the file Hello.js is never loaded in the build at all.

What do you mean? We simplify your module to `"HI"` string, because you always return `"HI"`.

> Apologies, you are correct, it's working fine actually.

If you provide more complex example or example of your problem I can help you.

**Comment by MartinDawson** [comment]
I figured out what is happening:

Here's the `entry` in my webpack.config.cjs:

Webpack:
```js
  entry: ["./app.js"],
```

However my workers where initialized in a file that is not the app.js but in this files imports:

app.js

```js
import api from "./src/api";

```
api.js

```js
const dcfModelWorker = new Worker(
  new URL("./workers/dcfModel.worker.js", import.meta.url),
  {
    type: "module",
  },
);
```

However if we change the above to this:

app.js

```js
import api from "./src/api";

const dcfModelWorker = new Worker(
  new URL("./src/workers/dcfModel.worker.js", import.meta.url),
  {
    type: "module",
  },
);

```
Then the bundling and chunking of the workers happens correctly.

So I think this is a bug.

If the workers are not in the entry files they won't be bundled properly and the worker imports don't get bundled or chunked.

**Comment by alexander-akait** [comment]
I don't see any problems here

**Comment by alexander-akait** [comment]
Please provide full example of problem(s) and describe it

**Comment by MartinDawson** [comment]
@alexander-akait Sorry but I still have trouble on a different issue with the workers.

Is their any way to provide configuration to webpack for workers?

The way the bundling of workers happens is that it seems to look for a `new Worker()` constructor. However if we use a library such as `piscina`: https://www.npmjs.com/package/piscina that wraps this call internally then webpack does not bundle it.

I know the old worker-plugin scanned for static strings instead of constructors to do this.

**Comment by alexander-akait** [comment]
`new Worker()` is just new endpoint (`dynamic`) with special `target` and some features

> The way the bundling of workers happens is that it seems to look for a new Worker() constructor. However if we use a library such as piscina: https://www.npmjs.com/package/piscina that wraps this call internally then webpack does not bundle it.

Can you provide example? I think problem here `path_1.resolve(__dirname, 'worker.js')`

**Comment by sokra** [comment]
> The way the bundling of workers happens is that it seems to look for a `new Worker()` constructor. However if we use a library such as `piscina`: [npmjs.com/package/piscina](https://www.npmjs.com/package/piscina) that wraps this call internally then webpack does not bundle it.

Yeah we are aware of this limitation. To make it as automatic as possible the syntax used has to statically analyse that a Worker is used. In most cases it's easy to change libraries to work with statically analysable syntax.

e. g. 

``` js
const piscina = new Piscina({
  factory: () => new Worker(new URL('./worker.mjs', import.meta.url))
});
```

**Comment by MartinDawson** [comment]
@sokra Thank you! That's what I was looking for.

Please can this be added to the documentation here: https://webpack.js.org/guides/web-workers/

It's definitely important to note it.
